chromadb/test/property/test_segment_manager.py

```diff
@@ -0,0 +1,125 @@
+import uuid
+
+import pytest
+import chromadb.test.property.strategies as strategies
+from unittest.mock import patch
+from dataclasses import asdict
+import random
+from hypothesis.stateful import (
+    Bundle,
+    RuleBasedStateMachine,
+    rule,
+    initialize,
+    multiple,
+    precondition,
+    invariant,
+    run_state_machine_as_test,
+    MultipleResults,
+)
+from typing import Dict
+from chromadb.segment import (
+    VectorReader
+)
+from chromadb.segment import SegmentManager
+
+from chromadb.segment.impl.manager.local import LocalSegmentManager
+from chromadb.types import SegmentScope
+from chromadb.db.system import SysDB
+from chromadb.config import System, get_class
+
+memory_limit = 100
+
+
+# Helper class to keep tract of the last use id
+class LastUse:
+    def __init__(self, n: int):
+        self.n = n
+        self.store = []
+
+    def add(self, id: uuid.UUID):
+        if id in self.store:
+            self.store.remove(id)
+            self.store.append(id)
+        else:
+            self.store.append(id)
+            while len(self.store) > self.n:
+                self.store.pop(0)
+        return self.store
+
+    def reset(self):
+        self.store = []
+
+
+class SegmentManagerStateMachine(RuleBasedStateMachine):
+    collections: Bundle[strategies.Collection]
+    collections = Bundle("collections")
+    collection_size_store: Dict[uuid.UUID, int] = {}
+    segment_collection: Dict[uuid.UUID, uuid.UUID] = {}
+
+    def __init__(self, system: System):
+        super().__init__()
+        self.segment_manager = system.require(SegmentManager)
+        self.segment_manager.start()
+        self.segment_manager.reset_state()
+        self.last_use = LastUse(n=40)
+        self.collection_created_counter = 0
+        self.sysdb = system.require(SysDB)
+        self.system = system
+
+    @invariant()
+    def last_queried_segments_should_be_in_cache(self):
+        cache_sum = 0
+        index = 0
+        for id in reversed(self.last_use.store):
+            cache_sum += self.collection_size_store[id]
+            if cache_sum >= memory_limit and index is not 0:
+                break
+            assert self.segment_manager._segment_cache[id][SegmentScope.VECTOR] is not None
+            index += 1
+
+    @invariant()
+    @precondition(lambda self: self.system.settings.is_persistent is True)
+    def cache_should_not_be_bigger_than_settings(self):
```

Comment:
whats the behavior for boundary conditions? Eg. if the limit is 10GB and we have two files - 6GB and 7GB, will we always only allow one?

Reply:
Yes, it this what we want? We could add a message in log when a collection got evicted for memory constraint

Reply:
Yeah lets log